src/gt4py/next/ffront/decorator.py

```diff
@@ -351,6 +359,61 @@ def _column_axis(self):
         return iter(scanops_per_axis.keys()).__next__()
 
 
+@dataclasses.dataclass(frozen=True)
+class ProgramWithBoundArgs(Program):
+    bound_args: dict[str, typing.Union[float, int, bool]] = None
+
+    def _process_args(self, args: tuple, kwargs: dict):
+        args = list(args)
+        b_args = list(self.bound_args.keys()) + args
+        param_ids = [param.id for param in self.past_node.params]
+
+        for name in self.bound_args.keys():
+            if name in kwargs:
+                # TODO(tehrengruber): use error with source location
+                raise RuntimeError(f"Parameter `{name}` already set as a bound argument.")
+
+        inp_args_len = len(b_args) - len(kwargs)
+        extra_args = len(param_ids) - inp_args_len
+        if extra_args > 0:
+            raise RuntimeError(
+                f"{extra_args} parameter(s) missing in new program call compared to original signature"
+            )
+
+        if inp_args_len != len(self.past_node.params):
+            raise RuntimeError(
+                f"Total number of arguments and keyword arguments ({inp_args_len}) does not match original program definition ({len(self.past_node.params)})!"
+            )
+
+        for index, param in enumerate(self.past_node.params):
+            if param.id in self.bound_args.keys():
+                args.insert(index, self.bound_args[param.id])
+
+        return super()._process_args(tuple(args), kwargs)
+
+    @functools.cached_property
+    def itir(self):
+        new_itir = super().itir
+        for new_clos_i, new_clos in enumerate(new_itir.closures):
+            for key in self.bound_args.keys():
+                index = next(
+                    index
+                    for index, closure_input in enumerate(new_clos.inputs)
+                    if closure_input.id == key
+                )
+                new_clos.inputs.pop(index)
+            new_args = [ref(inp.id) for inp in new_clos.inputs]
+            params = [sym(inp.id) for inp in new_clos.inputs]
+            for value in self.bound_args.values():
+                new_args.append(promote_to_const_iterator(literal_from_value(value)))
+            expr = itir.FunCall(
+                fun=itir.SymRef(id=itir.SymbolRef(list(self.closure_vars)[new_clos_i])),
```

Comment:
```suggestion
                fun=new_clos.stencil,
```
Getting the name of the stencil from `closure_vars` is not a good approach, something like this should work.